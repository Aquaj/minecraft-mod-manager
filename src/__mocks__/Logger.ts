import { vi } from 'vitest';

export const Logger = vi.importMock('../lib/Logger.js');
