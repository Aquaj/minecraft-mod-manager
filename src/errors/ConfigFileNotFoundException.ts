export class ConfigFileNotFoundException extends Error {
  constructor(configPath: string) {
    super(`Config file not found at ${configPath}`);
  }
}
