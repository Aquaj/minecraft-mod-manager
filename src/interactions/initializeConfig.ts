import { DefaultOptions } from '../mmm.js';
import { Loader, ModsJson, Platform, ReleaseType } from '../lib/modlist.types.js';
import inquirer from 'inquirer';
import { fileExists, writeConfigFile } from '../lib/config.js';
import * as path from 'path';
import { getLatestMinecraftVersion, verifyMinecraftVersion } from '../lib/minecraftVersionVerifier.js';

interface InitializeOptions extends DefaultOptions {
  loader?: Loader,
  gameVersion?: string,
  allowVersionFallback?: boolean,
  defaultAllowedReleaseTypes?: string,
  modsFolder?: string
}

interface IQInternal extends ModsJson {
  config: string
}

const addNewToFilename = (filename: string) => {
  const ext = path.extname(filename);
  const basename = path.basename(filename, ext);
  return `${basename}-new${ext}`;
};

const validateConfigName = () => async (input: string) => {
  if (await fileExists(input)) {
    return 'The config file already exists. Please choose a different name';
  }
  return true;
};

const validateGameVersion = () => async (input: string) => {
  if (await verifyMinecraftVersion(input)) {
    return true;
  }
  return 'Invalid game version, please try again';
};

const mergeOptions = (options: InitializeOptions, iq: IQInternal) => {
  return {
    loader: iq.loader || options.loader,
    gameVersion: iq.gameVersion || options.gameVersion,
    allowVersionFallback: iq.allowVersionFallback || options.allowVersionFallback,
    defaultAllowedReleaseTypes: iq.defaultAllowedReleaseTypes || options.defaultAllowedReleaseTypes!.replace(/\s/g, '').split(','),
    modsFolder: iq.modsFolder || options.modsFolder,
    mods: []
  };
};

export const initializeConfig = async (options: InitializeOptions, cwd: string): Promise<ModsJson> => {
  const prompts = [
    {
      when: await fileExists(options.config),
      type: 'confirm',
      name: 'overwrite',
      default: false,
      message: `The config file: (${options.config}) already exists. Should we overwrite it?`
    },
    {
      when: async (answers: { overwrite?: boolean }) => {
        return await fileExists(options.config) && !answers.overwrite;
      },
      type: 'input',
      name: 'config',
      default: addNewToFilename(options.config),
      message: 'What should we name the new config file?:',
      verify: validateConfigName
    },
    {
      when: !options.loader,
      default: options.loader,
      name: 'loader',
      type: 'list',
      message: 'Which loader would you like to use?',
      choices: Object.values(Platform)
    },
    {
      when: !options.gameVersion,
      name: 'gameVersion',
      type: 'input',
      default: await getLatestMinecraftVersion(),
      message: 'What exact Minecraft version are you using? (eg: 1.18.2, 1.19, 1.19.1)',
      validate: validateGameVersion
    },
    {
      when: !options.allowVersionFallback,
      name: 'allowVersionFallback',
      type: 'confirm',
      message: 'Should we try to download mods for previous Minecraft versions if they do not exists for your Minecraft Version?'
    },
    {
      when: !options.defaultAllowedReleaseTypes,
      name: 'defaultAllowedReleaseTypes',
      type: 'checkbox',
      choices: Object.values(ReleaseType),
      default: [ReleaseType.RELEASE, ReleaseType.BETA],
      message: 'Which types of releases would you like to consider to download?'
    },
    {
      when: !options.modsFolder,
      name: 'modsFolder',
      type: 'input',
      default: options.modsFolder || './mods',
      message: `where is your mods folder? (full or relative path from ${cwd}):`,
      validate: async (input: string) => {
        const dir = path.resolve(cwd, input);
        if (!await fileExists(dir)) {
          return `The folder: ${dir} does not exist. Please enter a valid one and try again.`;
        }
        return true;
      }
    }
  ];

  const iq = await inquirer.prompt(prompts) as IQInternal;

  const answers = mergeOptions(options, iq) as ModsJson;

  const configLocation = path.resolve(cwd, iq.config || options.config);

  await writeConfigFile(answers, configLocation);
  //TODO handle error

  return answers as ModsJson;
};

