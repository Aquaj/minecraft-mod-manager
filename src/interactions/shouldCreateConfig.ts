import inquirer from 'inquirer';

export const shouldCreateConfig = async (configLocation: string): Promise<boolean> => {
  const answers = await inquirer.prompt([
    {
      type: 'confirm',
      name: 'create',
      default: false,
      message: `The config file: (${configLocation}) does not exist. Should we create it?`
    }
  ]);
  return answers.create;
};
