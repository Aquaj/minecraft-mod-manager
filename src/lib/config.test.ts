import { beforeEach, describe, expect, it, vi } from 'vitest';
import fs from 'node:fs/promises';
import { chance } from 'jest-chance';
import {
  ensureConfiguration,
  fileExists,
  initializeConfigFile,
  readConfigFile,
  readLockFile,
  writeConfigFile,
  writeLockFile
} from './config.js';
import { ModInstall, ModsJson } from './modlist.types.js';
import path from 'node:path';
import { generateModInstall } from '../../test/modInstallGenerator.js';
import { generateModsJson } from '../../test/modlistGenerator.js';
import { ConfigFileNotFoundException } from '../errors/ConfigFileNotFoundException.js';
import { initializeConfig } from '../interactions/initializeConfig.js';
import * as process from 'process';
import { shouldCreateConfig } from '../interactions/shouldCreateConfig.js';
import { Logger } from './Logger.js';

vi.mock('../interactions/shouldCreateConfig.js');
vi.mock('../interactions/initializeConfig.js');
vi.mock('node:fs/promises');

describe('The config library', () => {
  let logger: Logger;
  beforeEach(() => {
    logger = new Logger({} as never);
    vi.resetAllMocks();
  });

  it('can determine that a file exists', async () => {
    vi.mocked(fs.access).mockResolvedValueOnce();
    expect(await fileExists(chance.word())).toBeTruthy();
  });

  it('can determine that a file does not exist', async () => {
    vi.mocked(fs.access).mockRejectedValueOnce(new Error());
    expect(await fileExists(chance.word())).toBeFalsy();
  });

  it('can write the config file', async () => {
    const config = {
      something: 'value'
    } as unknown as ModsJson;
    const configPath = path.resolve(chance.word());

    await writeConfigFile(config, configPath);

    expect(vi.mocked(fs.writeFile)).toHaveBeenCalledWith(
      configPath,
      '{\n'
      + '  "something": "value"\n'
      + '}'
    );
  });

  it('can write the lock file', async () => {
    const config = [{
      something: 'value'
    }] as unknown as ModInstall[];
    const configFileName = chance.word();
    const configPath = path.resolve(`${configFileName}.json`);
    const expectedLockFilePath = path.resolve(`${configFileName}-lock.json`);

    await writeLockFile(config, configPath);

    expect(vi.mocked(fs.writeFile)).toHaveBeenCalledWith(
      expectedLockFilePath,
      '[\n'
      + '  {\n'
      + '    "something": "value"\n'
      + '  }\n'
      + ']'
    );
  });

  it('can read the lock file when it exists', async () => {
    const configName = 'config.json';
    const lockfileName = 'config-lock.json';

    const randomModInstall = generateModInstall();

    const lockfileContents = JSON.stringify([randomModInstall.generated], null, 2);

    // Make file exist
    vi.mocked(fs.access).mockResolvedValueOnce();
    // Return file contents
    vi.mocked(fs.readFile).mockResolvedValueOnce(lockfileContents);

    const actualOutput = await readLockFile(configName);

    expect(actualOutput).toEqual([randomModInstall.expected]);

    expect(vi.mocked(fs.readFile)).toHaveBeenCalledWith(
      path.resolve(lockfileName),
      { encoding: 'utf8' }
    );

  });

  it('can returns an empty array and creates the file when the lock file does not exist', async () => {
    const configName = 'config.json';

    // Make file exist
    vi.mocked(fs.access).mockRejectedValueOnce(new Error());

    const actualOutput = await readLockFile(configName);

    expect(actualOutput).toEqual([]);

    expect(vi.mocked(fs.readFile)).not.toHaveBeenCalled();
    expect(vi.mocked(fs.writeFile)).toHaveBeenCalledWith(
      path.resolve('config-lock.json'),
      '[]'
    );

  });

  it('can read from the config file when it exists', async () => {
    const configName = 'config.json';

    // File exists
    vi.mocked(fs.access).mockResolvedValueOnce();

    const randomModsJson = generateModsJson();
    const fileContents = JSON.stringify(randomModsJson.generated, null, 2);

    // Return config file contents
    vi.mocked(fs.readFile).mockResolvedValueOnce(fileContents);

    const actualOutput = await readConfigFile(configName);

    expect(actualOutput).toEqual(randomModsJson.expected);
  });

  it('throws an error when the config file does not exist', async () => {
    const configName = path.resolve('config.json');

    // File does not exist
    vi.mocked(fs.access).mockRejectedValueOnce(new Error());

    await expect(readConfigFile(configName)).rejects.toThrow(new ConfigFileNotFoundException(configName));
  });

  it('can initialize a new config file', async () => {
    const configName = 'config.json';

    const randomConfig = generateModsJson().generated;

    vi.mocked(initializeConfig).mockResolvedValueOnce(randomConfig);
    const actual = await initializeConfigFile(configName, logger);

    expect(actual).toEqual(randomConfig);
    expect(vi.mocked(initializeConfig)).toHaveBeenCalledWith({ config: configName }, process.cwd(), logger);

  });

  describe('when ensuring that a config exists', () => {
    describe('and in interactive mode', () => {
      it('creates a new one if there is no existing one', async () => {
        const configName = 'config.json';
        const randomConfig = generateModsJson().generated;

        //File does not exist
        vi.mocked(fs.access).mockRejectedValueOnce(new Error());
        vi.mocked(shouldCreateConfig).mockResolvedValueOnce(true);
        vi.mocked(initializeConfig).mockResolvedValueOnce(randomConfig);

        const actual = await ensureConfiguration(configName, logger);

        expect(actual).toEqual(randomConfig);
        expect(initializeConfig).toHaveBeenCalledOnce();
      });

      it('throws an error if the user chose to not create one', async () => {
        const configName = 'config.json';

        //File does not exist
        vi.mocked(fs.access).mockRejectedValueOnce(new Error());
        vi.mocked(shouldCreateConfig).mockResolvedValueOnce(false);

        await expect(ensureConfiguration(configName, logger)).rejects.toThrow(ConfigFileNotFoundException);

        expect(initializeConfig).not.toHaveBeenCalled();
      });
    });

    describe('and in non-interactive mode', () => {
      it('throws an error', async () => {
        const configName = 'config.json';
        const quiet = true;

        //File does not exist
        vi.mocked(fs.access).mockRejectedValueOnce(new Error());

        await expect(ensureConfiguration(configName, logger, quiet)).rejects.toThrow(ConfigFileNotFoundException);
      });
    });

    it('returns the existing one if there is one', async () => {
      const configName = 'config.json';

      // File exists
      vi.mocked(fs.access).mockResolvedValueOnce();

      const randomModsJson = generateModsJson();
      const fileContents = JSON.stringify(randomModsJson.generated, null, 2);

      // Return config file contents
      vi.mocked(fs.readFile).mockResolvedValueOnce(fileContents);

      const actualOutput = await ensureConfiguration(configName, logger);

      expect(actualOutput).toEqual(randomModsJson.expected);
    });
  });
});
