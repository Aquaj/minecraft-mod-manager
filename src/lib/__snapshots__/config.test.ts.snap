// Vitest Snapshot v1

exports[`The config library > can initialize a new config file 1`] = `
{
  "allowVersionFallback": true,
  "defaultAllowedReleaseTypes": [
    "release",
    "beta",
  ],
  "gameVersion": "1.19.2",
  "loader": "fabric",
  "mods": [],
  "modsFolder": "mods",
}
`;
